call to update on already computed LazyRef

## Minimized code

This happens with 2 files in the same compilation unit and seems to be dependent on compilation order: if file2 comes first in lexicographical order it doesn't trigger. If it's all in 1 file, it doesn't trigger. If we compile file1, then add file2, then incrementally compile, it doesn't trigger. Without the `import` somewhat conflicting with the `val`, it doesn't trigger.

`scalaVersion := "0.24.0-RC1"`
`addSbtPlugin("ch.epfl.lamp" % "sbt-dotty" % "0.4.1")`

`file1.scala`
```Scala
object impl:
  case object UNone

import impl._

opaque type UOption[+A] = A | UNone.type
```

`file2.scala`
```Scala
val UNone: UOption[Nothing] = ???
```

[gist](https://gist.github.com/lourkeur/75bbb25c96e7a1181573f748161e04f4) (doesn't have plugins.sbt)

## Output (click arrow to expand)

<details>

```scala
[info] Executing in batch mode. For better performance use sbt's shell
[info] Compiling 2 Scala sources to /tmp/75bbb25c96e7a1181573f748161e04f4/target/scala-0.24/classes ...
[info] exception occurred while typechecking /tmp/75bbb25c96e7a1181573f748161e04f4/file1.scala
[info] exception occurred while compiling /tmp/75bbb25c96e7a1181573f748161e04f4/file1.scala, /tmp/75bbb25c96e7a1181573f748161e04f4/file2.scala
java.lang.AssertionError: assertion failed while compiling /tmp/75bbb25c96e7a1181573f748161e04f4/file1.scala, /tmp/75bbb25c96e7a1181573f748161e04f4/file2.scala
[error] ## Exception when compiling 2 sources to /tmp/75bbb25c96e7a1181573f748161e04f4/target/scala-0.24/classes
[error] java.lang.AssertionError: assertion failed
[error] dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
[error] dotty.tools.dotc.core.Types$LazyRef.update(Types.scala:2636)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.recur$3(SymDenotations.scala:493)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.setAlias$1(SymDenotations.scala:496)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.opaqueToBounds(SymDenotations.scala:510)
[error] dotty.tools.dotc.typer.Namer$TypeDefCompleter.opaqueToBounds$1(Namer.scala:1005)
[error] dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:1010)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:930)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:841)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:259)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:186)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:188)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1171)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1143)
[error] dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:133)
[error] dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:1928)
[error] dotty.tools.dotc.core.Types$Type.go$1(Types.scala:582)
[error] dotty.tools.dotc.core.Types$Type.findMember(Types.scala:750)
[error] dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:565)
[error] dotty.tools.dotc.core.Types$Type.member(Types.scala:549)
[error] dotty.tools.dotc.typer.Typer.memberInSelfButNotThis$2(Typer.scala:1979)
[error] dotty.tools.dotc.typer.Typer.classExistsOnSelf$3$$anonfun$2(Typer.scala:1981)
[error] dotty.tools.dotc.typer.Typer.classExistsOnSelf$4$$anonfun$adapted$1(Typer.scala:1986)
[error] scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:638)
[error] scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:634)
[error] scala.collection.AbstractIterator.foldLeft(Iterator.scala:1279)
[error] dotty.tools.dotc.typer.Typer.classExistsOnSelf$1(Typer.scala:1986)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1997)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2333)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2337)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2403)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2443)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2452)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2455)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2477)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2522)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2129)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2376)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2404)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2443)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2452)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2455)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2566)
[error] dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:42)
[error] dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error] dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:114)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:114)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:165)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:175)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:183)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:190)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:127)
[error] dotty.tools.dotc.Run.compile(Run.scala:110)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
[error] dotty.tools.dotc.Driver.process(Driver.scala:194)
[error] dotty.tools.dotc.Main.process(Main.scala)
[error] xsbt.CachedCompilerImpl.run(CachedCompilerImpl.java:69)
[error] xsbt.CompilerInterface.run(CompilerInterface.java:41)
[error] sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] java.lang.reflect.Method.invoke(Method.java:498)
[error] sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:248)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:122)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:95)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:91)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:186)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3(MixedAnalyzingCompiler.scala:82)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3$adapted(MixedAnalyzingCompiler.scala:77)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:215)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:77)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:146)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:343)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:343)
[error] sbt.internal.inc.Incremental$.doCompile(Incremental.scala:120)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:100)
[error] sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:180)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:98)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:102)
[error] sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:155)
[error] sbt.internal.inc.Incremental$.compile(Incremental.scala:92)
[error] sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:75)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:348)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:301)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:168)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:248)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:74)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:1762)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:1735)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:67)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:281)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error] sbt.Execute.work(Execute.scala:290)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:281)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
[error]            
[error] java.lang.AssertionError: assertion failed
[error]         at dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
[error]         at dotty.tools.dotc.core.Types$LazyRef.update(Types.scala:2636)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.recur$3(SymDenotations.scala:493)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.setAlias$1(SymDenotations.scala:496)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.opaqueToBounds(SymDenotations.scala:510)
[error]         at dotty.tools.dotc.typer.Namer$TypeDefCompleter.opaqueToBounds$1(Namer.scala:1005)
[error]         at dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:1010)
[error]         at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:930)
[error]         at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:841)
[error]         at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:259)
[error]         at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:186)
[error]         at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:188)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1171)
[error]         at dotty.tools.dotc.core.Denotations$SingleDenotation.computeAsSeenFrom(Denotations.scala:1143)
[error]         at dotty.tools.dotc.core.Denotations$PreDenotation.asSeenFrom(Denotations.scala:133)
[error]         at dotty.tools.dotc.core.SymDenotations$ClassDenotation.findMember(SymDenotations.scala:1928)
[error]         at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:582)
[error]         at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:750)
[error]         at dotty.tools.dotc.core.Types$Type.memberBasedOnFlags(Types.scala:565)
[error]         at dotty.tools.dotc.core.Types$Type.member(Types.scala:549)
[error]         at dotty.tools.dotc.typer.Typer.memberInSelfButNotThis$2(Typer.scala:1979)
[error]         at dotty.tools.dotc.typer.Typer.classExistsOnSelf$3$$anonfun$2(Typer.scala:1981)
[error]         at dotty.tools.dotc.typer.Typer.classExistsOnSelf$4$$anonfun$adapted$1(Typer.scala:1986)
[error]         at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:638)
[error]         at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:634)
[error]         at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1279)
[error]         at dotty.tools.dotc.typer.Typer.classExistsOnSelf$1(Typer.scala:1986)
[error]         at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1997)
[error]         at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2333)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2337)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2403)
[error]         at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2443)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2452)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2455)
[error]         at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2477)
[error]         at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2522)
[error]         at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2129)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2376)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2404)
[error]         at dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2443)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2452)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2455)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2566)
[error]         at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
[error]         at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
[error]         at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error]         at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:42)
[error]         at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
[error]         at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:114)
[error]         at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.immutable.List.foreach(List.scala:333)
[error]         at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:114)
[error]         at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:165)
[error]         at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]         at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]         at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]         at dotty.tools.dotc.Run.runPhases$5(Run.scala:175)
[error]         at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:183)
[error]         at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error]         at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
[error]         at dotty.tools.dotc.Run.compileUnits(Run.scala:190)
[error]         at dotty.tools.dotc.Run.compileSources(Run.scala:127)
[error]         at dotty.tools.dotc.Run.compile(Run.scala:110)
[error]         at dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
[error]         at dotty.tools.dotc.Driver.process(Driver.scala:194)
[error]         at dotty.tools.dotc.Main.process(Main.scala)
[error]         at xsbt.CachedCompilerImpl.run(CachedCompilerImpl.java:69)
[error]         at xsbt.CompilerInterface.run(CompilerInterface.java:41)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
[error]         at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:248)
[error]         at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:122)
[error]         at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:95)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:91)
[error]         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:186)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3(MixedAnalyzingCompiler.scala:82)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3$adapted(MixedAnalyzingCompiler.scala:77)
[error]         at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:215)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:77)
[error]         at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:146)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:343)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:343)
[error]         at sbt.internal.inc.Incremental$.doCompile(Incremental.scala:120)
[error]         at sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:100)
[error]         at sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:180)
[error]         at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:98)
[error]         at sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:102)
[error]         at sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:155)
[error]         at sbt.internal.inc.Incremental$.compile(Incremental.scala:92)
[error]         at sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:75)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:348)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:301)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:168)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:248)
[error]         at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:74)
[error]         at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:1762)
[error]         at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:1735)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:67)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:281)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error]         at sbt.Execute.work(Execute.scala:290)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:281)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]         at java.lang.Thread.run(Thread.java:748)
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed

```
</details>

--- a/compiler/src/dotty/tools/dotc/core/SymDenotations.scala
+++ b/compiler/src/dotty/tools/dotc/core/SymDenotations.scala
@@ -490,7 +490,10 @@ object SymDenotations {
         def recur(self: Type): Unit = self match
           case RefinedType(parent, name, rinfo) => rinfo match
             case TypeAlias(lzy: LazyRef) if name == this.name =>
-              lzy.update(tp)
+              if !lzy.completed then
+                lzy.update(tp)
+              else
+                throw CyclicReference(this)
             case _ =>
               recur(parent)
         recur(owner.asClass.givenSelfType)

--- a/tests/neg/i9052/A.scala
+++ b/tests/neg/i9052/A.scala
@@ -1,0 +1,6 @@
+object impl:
+  case object UNone
+
+import impl._
+
+opaque type UOption[+A] = (A | UNone.type) // error: Cyclic Reference involving UOption

--- a/tests/neg/i9052/B.scala
+++ b/tests/neg/i9052/B.scala
@@ -1,0 +1,1 @@
+val UNone: UOption[Nothing] = ??? // error: Cyclic Reference involving UOption

